### ISSUE TITLE
Support Custom Heading Elements for Modal

### ISSUE THREAD

**Issue by jsievenpiper** [Workday/canvas-kit#493]
## 🚀 Feature Proposal
By default, `Modal` seems to render the header provided to it inside an `h3`. In some cases, we have complex headings that would benefit from richer semantic markup for accessibility purposes. (Typically we'd probably use some pairing of a heading + subheading, but there are other uses, too.)

It would be great if `Modal`, as part of CKR4, understood the non-string case to be a _totally custom_ header, and wrapped the content in non-semantic markup, like a `div` or whatnot 🙂 

## Motivation
Currently, complex headers are not as accessible as they could be, as any header provided will be rendered as a child of an `h3`.

## Example
```tsx
const Header = () => (
  <>
    <h3>My Header</h3>
    <h4>For Item 43</h4>
  </>
)

const Header2 = () => (
  <>
    <SystemIcon icon={/* ... */} />
    <h3>My Header</h3>
  </>
)

<Modal header={Header} /* ... */ />
```

**Comment by mathalete** [comment]
Commenting as @NicholasBoll asked me to have a quick look at this.

In terms of accessibility best practice, the first visible heading in a dialog should be an `<h1>` as dialogs are considered to act as a separate document to the main page.

Thus any descendant headings within the modal would need to be logically nested and heading levels wouldn't be skipped.

Example

```
<div role="dialog" aria-modal="true" aria-labelledby="h1">
  <h1 id="h1">Dialog title</h1>
  ...
  <h2>Another heading</h2>
  ...
</div>
```
@jsievenpiper I'm not sure what you're proposing in terms of markup for complex headings - could you provide a couple of simple HTML examples? Or samples in the wild that you wish added to Canvas Kit?

**Comment by jsievenpiper** [comment]
Sorry it's taken me so long to get back to this... 

I haven't followed this in a post 4.0 world, so this may have changed. Effectively, no matter what content was provided to `Modal`'s `header` prop, that content was rendered inside an `h3` by canvas:

```tsx
const headerContent = `blah`
const modal = <Modal header={headerContent} ... />

// resulting structure:
// <div role="dialog"...>
//   <h3>blah</h3>
```
So for a simple example like this -- there's no problem at all. It's not the `h3` itself that is the problem (canvas could be using an `h1` or whatever is most accessible), the problem is when you want to be able to construct more complex headers that require complex styling.

This is fine if you have simple text content or text and presentational content only, but falls apart in the examples listed above. In this second example, I can recreate a pretty common Workday Design where my modal has an icon in the header to make it look spiffy.

```tsx
const headerContent = (
  <>
    <SystemIcon ... />
    <h3>My Header</h3>
  </>
)
const modal = <Modal header={headerContent} ... />

// resulting structure:
// <div role="dialog"...>
//   <h3>
//      <icon stuff...>
//      <h3>My Header</h3>
//    </h3>
```

In this second case it's a bit unwieldy, but I can still pull it off if I change my internal code to replace `headerContent`'s `h3` with a `span` or something. So still not a problem.

The case we were trying to accomplish without hacks looked like this:

```tsx
const headerContent = (
  <>
    <SystemIcon ... />
    <h3>Modal Title</h3>
    <h4>Modal Subtitle</h4>
  </>
)
const modal = <Modal header={headerContent} ... />

// resulting structure:
// <div role="dialog"...>
//   <h3>
//      <icon stuff...>
//      <h3>Modal Title</h3>
//      <h4>Modal Subtitle</h4>
//    </h3>
```
Visually in this design, the `h3` and `h4` were inline with each other.

Now there's nothing we can do -- we're in the context of the `h3` (or again, whatever heading it may be), and we're not free to break out of that, so the nested h4 is semantically incorrect. If we try the same `span` hack as above, we lose our semantic heading levels. We _could_ skip the `Modal`'s `header` prop altogether, but then we're sort of breaking the `Modal`'s semantics itself by rendering our header as body content (which, is what we settled on, for what it's worth). It happened to work with CKR3, but relies on the internal details of the modal not changing.

It so happened that we had other (common but currently unsupported) requirements like fixed headers and footers in our modals that we more-or-less ended up rolling our own after all the customization, but if those things theoretically made it into CKR as standard features, this API would still block the team from being able to uptake.

I could see a couple plausible solutions:
- first class support for the designs being built internally (e.g. fixed icon, subheading support). Less flexible, but looking fabulous with no effort
- making it such that if the `header` prop is text, wrap it in a heading of some sort (`h3`, `h1`, whatever is considered best), but if the content is some sort of React node or fragment, pasting it directly in.

The second one obviously comes at the caveat of allowing teams to accidentally break accessibility if they're not sure what they are doing.

Hope that helps, thanks for looking into this!

**Comment by NicholasBoll** [comment]
This should now be fixed in version 5:

```tsx
<Modal>
  <Modal.Target as={DeleteButton}>Delete Item</Modal.Target>
  <Modal.Overlay>
    <Modal.Card>
      <Modal.CloseIcon aria-label="Close" />
      <Modal.Heading as="div">
        <SystemIcon aria-hidden="true" ... />
        <h3>Modal Title</h3>
        <h4>Modal Subtitle</h4>
      </Modal.Heading>
      <Modal.Body>
        ...
      </Modal.Body>
    </Modal.Card>
  </Modal.Overlay>
</Modal>
```

The `<Modal.Heading as="div">` is required for accessibility. The `Modal.Heading` adds an `aria-labelledby` which will name the Modal according to the text inside the `div`. If you desire to instead name by the "Modal Title" and not the "Modal Subtitle", do this:

```tsx
...
<Modal.Card>
  <Modal.CloseIcon aria-label="Close" />
  <SystemIcon aria-hidden="true" ... />
  <Modal.Heading as="h3">Modal Title</Modal.Heading>
  <h4>Modal Subtitle</h4>
</Modal.Heading>
<Modal.Body>
  ...
</Modal.Body>
```
